Mathlib/LinearAlgebra/ExteriorPower/Basic.lean

```diff
@@ -0,0 +1,427 @@
+/-
+Copyright (c) 2024 Sophie Morel. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sophie Morel
+-/
+import Mathlib.LinearAlgebra.ExteriorAlgebra.OfAlternating
+import Mathlib.LinearAlgebra.TensorPower
+
+/-!
+# Exterior powers
+
+We introduce the exterior powers of a module `M` over a commutative ring `R`.
+
+## Notation
+
+The `n`th exterior power of the `R`-module `M` is denoted by `ExteriorPower R n M`; it is a module
+over `R`, defined as `LinearMap.range (ExteriorAlgebra.ι R : M →ₗ[R] ExteriorAlgebra R M) ^ n`.
+We also introduce the notation `Λ[R]^n M` for `ExteriorPower R n M`.
+
+## Definitions
+
+* `ExteriorPower.ιMulti` is the canonical alternating map on `M` with values in `Λ[R]^n M`.
+
+* If `f` is a `R`-linear map from `M` to `N`, then `ExteriorPower.map n f` is the linear map
+between `n`th exterior powers induced by `f`.
+
+* `ExteriorPower.toTensorPower`: linear map from the `n`th exterior power to the `n`th
+tensor power (coming from `MultilinearMap.alternatization` via the universal property of
+exterior powers).
+
+## Theorems
+
+* The image of `ExteriorPower.ιMulti` spans `Λ[R]^n M`.
+
+* `ExteriorPower.liftAlternatingEquiv` (universal property of the `n`th exterior power of `M`):
+the linear equivalence between linear maps from `Λ[R]^n M` to a module `N` and `n`-fold
+alternating maps from `M` to `N`.
+
+* `ExteriorPower.map_injective_field`: If `f : M →ₗ[R] N` is injective and `R` is a field, then
+`ExteriorPower.map n f` is injective.
+
+* `ExteriorPower.map_surjective`: If `f : M →ₗ[R] N` is surjective, then `ExteriorPower.map n f`
+is surjective.
+
+* `ExteriorPower.mem_exteriorPower_is_mem_finite`: Every element of `Λ[R]^n M` is in the image of
+`Λ[R]^n P` for some finitely generated submodule `P` of `M`.
+
+-/
+
+open BigOperators
+open scoped TensorProduct
+
+universe u v uM uN uN' uN'' uE uF
+
+variable {R : Type u}{n : ℕ} {M : Type uM} {N : Type uN} {N' : Type uN'} {N'' : Type uN''}
+  [CommRing R] [AddCommGroup M] [Module R M] [AddCommGroup N] [Module R N] [AddCommGroup N']
+  [Module R N'] [AddCommGroup N''] [Module R N'']
+
+variable {K : Type v} {E : Type uE} {F: Type uF} [Field K] [AddCommGroup E] [Module K E]
+  [AddCommGroup F] [Module K F]
+
+variable (R M n)
+
+/--Definition of the `n`th exterior power of a `R`-module `M`. We introduce the notation
+`Λ[R]^n M` for `ExteriorPower R n M`. -/
+@[reducible]
+def ExteriorPower := (LinearMap.range (ExteriorAlgebra.ι R : M →ₗ[R] ExteriorAlgebra R M) ^ n)
+
+@[inherit_doc]
+notation:100 "Λ[" R "]^" n:arg => ExteriorPower R n
```

Comment:
I think `(LinearMap.range (ExteriorAlgebra.ι R : M →ₗ[R] ExteriorAlgebra R M) ^ n)` is already used quite a lot throughout mathlib; can you split out a PR that just adds this new `def` and notation, and replaces all the existing uses with the new notation?

Reply:
Can do. Given that this expression already appears in `LinearAlgebra/ExteriorAlgebra/Basic.lean`, I guess it would make sense to introduce `ExteriorPower` there already.

Reply:
It's maybe worth waiting to see what Zulip says about the type of this expression, Right now Lean is inferring
```suggestion
`Λ[R]^n M` for `ExteriorPower R n M`. -/
@[reducible]
def ExteriorPower : Submodule R M :=
  (LinearMap.range (ExteriorAlgebra.ι R : M →ₗ[R] ExteriorAlgebra R M) ^ n)

@[inherit_doc]
notation:100 "Λ[" R "]^" n:arg => ExteriorPower R n
```
but this goes against the naming convention, since `ExteriorPower` is not a type.

Either we should make it a type (in which case it won't actually work as a substitute in a handful of places), or we should rename it to `exteriorPower`.

Reply:
Right, I am actually using that it's a submodule of the exterior algebra in a few places. Anyway, I started a new branch that defined `ExteriorPower` earlier and started substituting it [there](https://github.com/leanprover-community/mathlib4/tree/SM.ExteriorPowerDef), and it's creating timeouts in `ExteriorAlgebra/Grading.lean`. I need to play around with it more to see if I can figure out what happened.

Reply:
Okay, I changed all the `ExteriorPower`s to `exteriorPower`s and created PR #10744 to introduce the definition and notation. (But it does have that annoying instance synthesization problem.)